pip/basecommand.py

```diff
@@ -314,6 +314,23 @@ def populate_requirement_set(requirement_set, args, options, finder,
                     'You must give at least one requirement to %(name)s '
                     '(see "pip help %(name)s")' % opts)
 
+        # On Windows, any operation modifying pip should be run as:
+        #     python -m pip ...
+        # See https://github.com/pypa/pip/issues/1299 for more discussion
+        should_show_use_python_msg = (
+            sys.platform == 'win32' and
+            requirement_set.has_requirement('pip') and
+            "pip" in os.path.basename(sys.argv[0])
+        )
+        if should_show_use_python_msg:
+            new_command = [
+                sys.executable, "-m", "pip"
+            ] + sys.argv[1:]
+            raise CommandError(
+                'To modify pip, please run the following command:\n{}'
+                .format(" ".join(new_command))
+            )
```

Comment:
Would this be better as a warning rather than an error? I can't immediately think of an existing scenario where the test ```should_show_use_python_msg``` would give a false positive, but I'd rather not block users from proceeding if they know what they are doing is OK.

(Hypothetical example, if there's a 3rd party tool called ```upgrade-pip.exe``` that uses the ```pip.main``` entry point to upgrade pip, it would be caught by this check even though it's perfectly OK. That's arguably OK as we don't support using pip's API directly, but it's still a bit unfriendly...)